Add hostname to Locks

nupplaphil opened the pull request:
This should help with #9095 and should work in a way @annando hopefully will like too :-)

We now add the hostname to each lock as well. The hostname is detected based on either `uname -n` or per environment variable `NODE_NAME`. This will have the effect that for normal setups, nothing changes except that in the db, there's now a new column "hostname" as well. But for cluster, we can now override this behaviour with "NODE_NAME" where each node can manually override this value with it's node-name-id .

And now the workers are ready for multi-nodes :-)

nupplaphil committed:
Add "getmypid()" for DB-Locking to factory instead of constructor

nupplaphil committed:
Add the real node/server "hostname" to the lock/cache mechanism instead of the hostname of the baseUrl (enables multi-server locking).

nupplaphil committed:
Add Hostname detection, tests & env-variable for cluster pod nodes

nupplaphil committed:
Add dependency to Node because of the $_SERVER variable

nupplaphil committed:
Add EOF

nupplaphil committed:
Make NodeTest more flexible

annando commented:
Looks great!

I'm currently thinking if it would be better having a "host" table, so that we could work with host id values. This would be better for the workerqueue table since it seen only would add an integer field and not huge text field. But this is something for the future.

This currently only works when you use the database for the locks, I guess?

nupplaphil committed:
Add possibility to use hostnames for cache locking too

nupplaphil commented:
:-)

I'll give the host-table a try :-)

The cache locking will work in a few minutes too, I'm adding the node hostname to the lock key as well so it makes it unique too.

AlfredSK commented:
DB version is already on 1366 now.

MrPetovan commented:
Thanks for the heads-up, I'm expecting more schema changes from @nupplaphil so he'll be able to rebase again latest develop.

annando commented:
@nupplaphil if it speeds up stuff I can implement a "host" model this evening. Then you only had to change your functions to take an integer value (I would suggest unsigned tinyint - 255 hosts ought to be enough for anybody)

annando commented:
@nupplaphil I just created a small host model class: https://github.com/friendica/friendica/pull/9120

It is lacking tests and checks.

nupplaphil commented:
thx @annando maybe I can check it today and finish this PR.

nupplaphil committed:
Make Host Model testable

nupplaphil committed:
Use host-id instead of hostname for locking

nupplaphil committed:
Remove unused DI::host()

nupplaphil committed:
Use zhe host id for the lock factory

MrPetovan commented:
Tests are failing.

nupplaphil commented:
yep, that's why I changed it to draft. I do make some test on my friendica node with different locks as well to make sure it all works as expected :-)

nupplaphil committed:
Fix tests

nupplaphil committed:
Rename table "host" to "hosts" & fix update routine

nupplaphil committed:
Fix update routines (check if column/tables exist)

nupplaphil commented:
There were some troubles with pre update routines. When we want to update from a really old version to a newer, the pre updates are possibly failing because of not existing tables/columns. We should cover every DB change in a pre update with a corresponding check. I added it for the recent updates. You can reproduce this error when using the Friendica db fixture of the root and call the update routine afterwards.

Another thing is that I do want to rename `host` to `hosts`, because DB tables are mainly plural an model classes singular (at least I found most of the tables following this pattern) and because I found tinyint really tiny, I changed it to mediumint.

All this changes are hopefully done before somebody is using the dev branch, otherwise I do need support how to add the specific logic in the update methods..

annando commented on database.sql:
```diff
@@ -588,12 +588,12 @@ CREATE TABLE IF NOT EXISTS `hook` (
-CREATE TABLE IF NOT EXISTS `host` (
-	`id` tinyint unsigned NOT NULL auto_increment COMMENT 'sequential ID',
+CREATE TABLE IF NOT EXISTS `hosts` (
+	`id` mediumint unsigned NOT NULL auto_increment COMMENT 'sequential ID',
```
`smallint` really is enough. This does cover 65,535 hostnames. `Mediumint` does cover 16,777,215 - which is much too much. This field will be used in tables like the workerqueue that is used in a high frequency, so I want to save as many space as possible.

MrPetovan replied:
`as many space` => `as much space`

`as much` needs to be used with innumerables (like space, weight, etc...) while `as many` needs to be used with numerables (like people, apples, etc...)

annando commented on static/dbstructure.config.php:
```diff
@@ -653,10 +653,10 @@
-	"host" => [
-		"comment" => "Hostname",
+	"hosts" => [
+		"comment" => "Host table",
```
We should comment what the table contains and not tell that is a table.

annando commented:
I strongly advocate against renaming the `host` table to `hosts`. By now our tables are always using the singular like "contact", "hook", "tag", "item", ... 

Also when reading the field data we fetch this data in a `$contacts` variable which is split into a `$contact` table for every row.

nupplaphil commented:
Ah ok... I know from our db experts that they advice/force us to use plural table names because a table contains bunch of same rows (plural) but a model is used for one specific row so singular. But i'll revert it again :-)

nupplaphil committed:
Reverted DB table hosts to host

nupplaphil commented on update.php:
```diff
@@ -739,12 +788,40 @@ function pre_update_1364()
+		!DBStructure::existsColumn('locks', ['host-id'])) {
+
+		// If the locks table doesn't have a host-id, but values, you have to temporary disable the foreign checks!
+		if (!DBA::e('SET FOREIGN_KEY_CHECKS=0')) {
```
@annando do you think does this work? So the dbstructure is able to create the foreign key constraints and in the `update_1367()` function, I set the right values and re-enable the foreign check?

annando replied:
I'm unsure about that. I also don't know what happens when the checks are enabled again and there is still a violation. That's why I delete all violations first.

nupplaphil commented:
the branch is running on my friendica node, I'll reopen it at least at sunday in case there aren't any errors :-)

annando commented:
Concerning singular and plural with tables: At our company there is the opposite rule that is also a rule set by Microsoft for our product. There the tables are always singular since we see them as a row, while the pages that display multiple lines of these records are named in plural.

annando commented:
Curious question: How is the model class called to get the id for a given hostname (or the system's hostname) returned?

nupplaphil commented:
Here's where we get the `Host` instance in the lock factory:
https://github.com/friendica/friendica/pull/9114/files#diff-8b66e92570cbb5b91e9c90f7f1d85732R71
Dice is therefore implicit calling the `Host` constructor befoee:
https://github.com/friendica/friendica/pull/9114/files#diff-1a0238e0b4b946b62014c5097a986c49R49
And here we're setting the host id for passing it to all the different lock mechanism:
https://github.com/friendica/friendica/pull/9114/files#diff-8b66e92570cbb5b91e9c90f7f1d85732R77

annando commented:
My question was more like: How do I call the model functions?

nupplaphil commented:
Ah ok. You need a `DI` method: `DI:host()->getId()` this is needed to mock the dba calls in the Host class for testing and to mock the host class per se for the lock tests

annando commented on src/Model/Host.php:
```diff
@@ -21,59 +21,91 @@
-	 * @param int $id
+	 * @return integer host name id
+	 */
+	public function getId()
```
Could you extend this function so that it can return the id for a given hostname as well?

MrPetovan replied:
What would be the use?

annando replied:
When there are several machines that are running the worker queues, they could be different in their performance. So you might want to limit the number of parallel worker tasks different per machine. So you would need a configuration for that by hostname.

MrPetovan replied:
Then we'd probably need a separate Repository class to retrieve a `Host` model by name/id, and detect and instantiate the local `Host` object. The `Model\Host` class would probably become an Active Record like `Model\Introduction`.

nupplaphil replied:
A model class is responsible for representing a specific host of the persistence layer. So for me the `getId()` is responsible for the concrete Id of this concrete host row. I don't want to add any way to override the inner state of this object. Things like deciding the amount of workers based on hosts is a responsibility of the method which is **using** the Host, but not for the host itself.

So I'd prefer having two different ways:
* A unique `Host` class, which will get called everytime "host" is retrieved during a complete call
  * per `DI::host()`
  * if automatically called per Dice inside a constructor (like for the constructor in the `LockFactory`)
* A `Repository\Host` class for selecting or altering host entries
  * per `DI::hostRepo()`
  * if automatically called per Dice inside a constructor (like `__constructor(Repository\Host $hostRepo)`)

annando replied:
Is `Repository` the new `Model`? Because until now I thought that `Model` is - for example - the interface between code and database for a specific table?

nupplaphil replied:
The `Repository` is the interface between a **whole table** and the code
The `Model` is the interface between **one or many rows** of a table and the code

annando commented on src/Model/Host.php:
```diff
@@ -21,59 +21,91 @@
-	 * @throws \Exception
 	 */
-	public static function getName(int $id)
+	public function getName()
```
Could you extend this function to return the host name of a given id as well?

MrPetovan replied:
What would be the use?

annando replied:
It could be interesting for the logging to see directly the values per hostname.

nupplaphil replied:
same as https://github.com/friendica/friendica/pull/9114#discussion_r484103828

annando commented on src/Model/Host.php:
```diff
@@ -21,59 +21,91 @@
+
+	private function detectHostname(Database $dba, array $server)
 	{
+		$hostname = $server[self::ENV_VARIABLE] ?? null;
```
When I understood it correctly, then this `server` variable is only used during the tests? If yes, then why is it an array and not simply a string?

MrPetovan replied:
No, in production the `$server` parameter of the constructor is provided by Dice according to the rules in `static/dependencies.config.php`.

annando replied:
Then I recommend reducing this to always use the appropriate PHP function and no environment variable. It adds some uncertainy. The CLI PHP could fetch its hostname via another way than the frontend PHP process. And since they then origin from a different source, we cannot guarantee for sure that the values on the same machine were identical.

Edit: Under https://www.php.net/manual/en/reserved.variables.server.php I haven't found "NODE_NAME" as an official value. And my machines aren't providing this variable.

MrPetovan replied:
What is the appropriate PHP function? In a containerized environment (like it is the target), the environment variable would be set automatically by the Docker script.

annando replied:
I meant the PHP function that is used when the variable isn't set.

MrPetovan replied:
The null-coalescing operator `??` is an appropriate PHP function when a variable isn't set. What did you have in mind?

annando replied:
No. I meant to use the PHP function to fetch the hostname all the time and not only when it isn't found in the `$_SERVER` array.

MrPetovan replied:
I don't think it's necessary since host names wouldn't be used outside of a containerized setup but maybe @nupplaphil can shed more light on this particular code decision.

annando replied:
Well, like I said: The mixture of having a frontend PHP process and a backend PHP process could lead to a situation where the hostnames are detected differently. And this could cause weird problems. I think that there is really no reason to not always use the PHP function to fetch the hostname.

MrPetovan replied:
Not with an environment variable that was set by Docker. It still would be the same (virtual) machine (and its environment variables) whether it's a frontend process or a backend process.

annando replied:
But this does have to run on regular systems as well.

MrPetovan replied:
On single-host nodes, the hosts table wouldn't matter so it's fine if this code returns null because the environment variable wouldn't be set.

annando replied:
This code wouldn't return `null` because it falls back on the PHP function for fetching the hostname.

nupplaphil replied:
`php_uname` is **the** way to retrieve the hostname in a "normal" PHP environment. But for cluster environments, we cannot trust the `hostname` entry of a server anymore. I found references where the cluster-manager sets **every** `/etc/hostname` to `localhost`. So the usage of `php_uname` is not adviced from cluster experts.

Instead they're using this method:
https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/#use-pod-fields-as-values-for-environment-variables , in particular this section:
```yml
      - name: MY_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
```
which is setting the `metadata.name` of a node (= the unique name of the node) as the envionment variable `MY_POD_NAME`. So I do need a way to disable `php_uname` and instead use a environment variable for it.

I thought doing it inside the `Host` constructor is a good idea. But if we really want a own "repository" class (see https://github.com/friendica/friendica/pull/9114#discussion_r484103828 ) I'm with you that the decision if using `php_uname` or the environment variable, should be done outside of the Host class itself.

annando replied:
I would suggest some configuration that defaults to `gethostname()` and only if configured, it should take the value from a different source. I guess that people who can set up these clusters can handle this.

nupplaphil replied:
There's now a "selectCurrentHost()" method in the repository, which either uses `gethostname()` (as suggested instead of the `php_uname('n')` method or checks the env-variable

annando commented on src/Model/Host.php:
```diff
@@ -21,59 +21,91 @@
+		$hostname = trim($hostname);
+		if (empty($hostname)) {
+			$this->logger->error('Empty hostname is invalid.');
+			$this->id = -1;
```
I think the system should return `0` for an empty hostname. This whould ensure that the other systems could process without any problem with storing the values into an unsigned integer field.

nupplaphil replied:
ok

annando commented on static/dependencies.config.php:
```diff
@@ -221,5 +221,10 @@
+	],
+	Friendica\Model\Host::class => [
+		'constructParams' => [
+			$_SERVER,
```
What is the `$_SERVER` variable doing here?

MrPetovan replied:
It is provided as the `$server` parameter of the `Model\Host` constructor by Dice when it is instantiated as part of a chained dependency. For example, calling `DI::lock()` orders Dice to create an instance of the `ILock` interface. But in PHP interfaces cannot be instantiated. Instead, we added a rule to `static/dependencies.config.php` to make Dice create a `LockFactory` instance when asked to create an `ILock` instance. Then Dice looks at the constructor of this class which is the following:
```php
public function __construct(CacheFactory $cacheFactory, IConfig $config, Database $dba, LoggerInterface $logger, Host $host)
```
Dice then proceeds to create or retrieve (if they was already created) instances of the `CacheFactory`, `IConfig`, `Database`, `Loggerinterface` and `Host` classes.

In the case of `Model\Host`, it cannot guess what it should provide for the `$server` parameter since it's an array and not a class that it could instantiate. So we have to provide the parameter value in the `static/dependencies.config.php`.

annando replied:
Okay. This whole construct is still quantum physics to me - while I'm stuck at newtonian physics.

MrPetovan replied:
I understand, it took me a little while to piece everything together.

nupplaphil replied:
Instead of using Dice, we could write this code, starting with the "first" dependency:
```php
$cache = new Core\Config\Cache();
$profiler = new Util\Profiler($cache);
/**  @var LoggerInterface */
$logger = new Util\Logger\VoidLogger();
$dba = new Database($cache, $profiler, $logger, $_SERVER);

//
// Here's the "new" magic:
//
$host = new Host($dba, $_SERVER);

/** @var ILock */
$lock = new DatabaseLock($dba, $host->getId(), getmypid());
```
That's how "dependency injection" works in detail. It looks at the constructor of the requested class and then it recursive checks the constructor of each dependency and their constructor until it finds the very first dependency without any other dependency (in Friendica, it's the `Core\Config\Cache`) and from this point, it creates every class. This system/pattern is called "inversion of control".

Why not per hand:
- It's repetitive the same and complex (I skipped every Factory-decision and the creation of the "real" Logger)
- Dice caches instances, so in case of a re-request, no re-creation is needed (this is configured with the `shared => true` parameter in `dependencies.config.php`)

Why do we even use this pattern?
- We can mock each dependency, passed per constructor (=> make it testable)
- It shows us the "dependency tree" and errors are instant thrown if there are dependency cycles (like `Core\Config\Cache` would have had dependencies to other classes)
- Each class, using DI, can be sure that every needed pre-dependency is computed and set right => separation of concern

But, to be honest, as @annando already notices, the disadvantage of this architectural pattern is that the complexity of the overall codebase increased (because of the implicit constructor).

nupplaphil committed:
Add Host Repository & add current host selection logic to it

fix wrong var name

nupplaphil committed:
fix wrong var name

nupplaphil committed:
Change update routine

nupplaphil committed:
Fix Test&Factory

nupplaphil commented:
finished :-)

nupplaphil commented on update.php:
```diff
@@ -739,12 +788,27 @@ function pre_update_1364()
+function pre_update_1368()
+{
+	// Lock cleanup to avoid constraint violations during foreign key creation
+	if (DBStructure::existsTable('locks') &&
```
@annando do you think, this works as expected? Or do we possibly brick something when deleting locks which are maybe in use?

annando replied:
The system **should** automatically go in maintenance during update. So nothing should be executed, so it should be no problem.

nupplaphil commented:
thought I got it .. But I experimented with my local vagrant environment and there's a problem left:
The `Update::run()` routine tries to add a lock, but the lock mechanism of the current codebase needs the host table, which isn't added yet ... So the update failed :-/

nupplaphil commented:
I switched branch (the rebase is open yet, so don't worry about the strange commits).
=> I think this enhancement is too big to be part of the RC phase. In particular the upgrade process can easily brick if not enough tested!

nupplaphil commented:
This one makes me headache .. I was far far away from this PR because of it...

- The Update is not possible because the Update needs a lock entry (the table we now try to update)
- So I started refactoring the Update class to inject alternative locking methods in case of updating the lock table itself
- But the Update class uses DBStructure which isn't a dependency so I cannot inject the DBStructure class
- So I started refactoring DBStructure
- But the DBStructure class is used very often so this would become a real big PR ..

@MrPetovan @annando do you have any ideas how I'm able to update the `lock` table itself within a automatic update process? Without a solution for this problem, the update because of this change would need a manual approach ..

MrPetovan commented:
No, we don't have an external system to update tables, it is deeply part of the existing system like you found out.

annando commented:
I geuss this is still WIP, so we should move it.

nupplaphil committed:
PHP CS fixings :-)

nupplaphil committed:
Add own update routine for locks&hosts

nupplaphil commented on src/Database/DBStructure.php:
```diff
@@ -425,6 +425,38 @@ public static function install(string $basePath)
+	 *
+	 * @throws Exception
+	 */
+	public static function updateLockTable(string $basePath, bool $verbose, bool $action, bool $install)
```
@annando I'm not 100% sure if this is an good idea ...

This function will be called **before** any locks are used during the update and should update/create the `host` and `locks` table beforehand (without any locking mechanism ...)

At my local node, I was able to upgrade from version 1290 (`friendica_test_data.sql`) to this version..

annando replied:
This sounds a little bit like the existing function `checkInitialValues`

nupplaphil replied:
But it's a structure upgrade, not an insert of values.. But yeah, it's similar I think

nupplaphil replied:
But the main question: does it work? :D .. For simple updates it will.. But are there circumstances where a structure update with an own dba lock before the main update lock could lead to problems? 

I think I could add another check if there's a update lock and stop the update ..

annando replied:
I'm not even sure if this does work at all.

nupplaphil committed:
Fix database.sql name
